tools/chpldoc/findUndocumentedSymbols.py

```diff
@@ -0,0 +1,362 @@
+    """
+    return (attrs := node.attribute_group()) and any(a.name() == marker for a in attrs)
+
+
+def is_deprecated(node: dyno.AstNode) -> bool:
+    """a node is deprecated if it or its parent has an attribute that is deprecated"""
+    return node_has_attribute(node, "deprecated")
+
+
+def is_unstable(node: dyno.AstNode) -> bool:
+    """a node is unstable if it or its parent has an attribute that is unstable"""
+    return node_has_attribute(node, "unstable")
+
+
+def is_nodoc(node: dyno.AstNode) -> bool:
+    """
+    a symbol is marked nodoc for the following reasons
+    - symbol has attribute `@chpldoc.nodoc`
+    - symbol name is prefixed with `chpl_` and symbol does not have `pragma "chpldoc ignore chpl prefix"`
+    - symbol is private
+    """
+    if node_has_attribute(node, "chpldoc.nodoc"):
+        return True
+
+    if (
+        hasattr(node, "name")
+        and node.name().startswith("chpl_")
+        and "chpldoc ignore chpl prefix" not in node.pragmas()
+    ):
+        return True
+
+    if hasattr(node, "visibility") and node.visibility() == "private":
+        return True
+
+    return False
+
+
+def is_docstring_comment(c: dyno.Comment) -> bool:
+    """c is a docstring if it doesn't begin with '//'"""
+    return not c.text().startswith("//")
+
+
+def get_node_name(node: dyno.AstNode) -> List[str]:
+    """
+    get the name of a node, even when it may not have a `name()` method
+
+    if this is a member of a aggregate type, also gets that name
+    if this is a member of a module, also gets that name
+
+    For a multidecl, gets all the names
+
+    """
+
+    def get_simple_node_names(node: dyno.AstNode) -> List[str]:
+        names: List[str] = []
+        if hasattr(node, "name"):
+            names.append(node.name())
+        elif isinstance(node, dyno.MultiDecl):
+            names.extend(c.name() for c in node)
+        else:
+            names.append(str(node))
+        return names
+
+    def get_single_name(node: dyno.AstNode) -> str:
+        names = get_simple_node_names(node)
+        if len(names) != 1:
+            print("Error: violated invariant")
+            exit(1)
+        return names[0]
+
+    names_to_return = []
+
+    base_names = get_simple_node_names(node)
+    for base_name in base_names:
+        name = base_name
+        # handles secondary methods
+        if (
+            hasattr(node, "this_formal")
+            and (this := node.this_formal())
+            and (typename := this.type_expression())
+        ):
+            aggregate_name = get_single_name(typename)
+            name = f"{aggregate_name}.{name}"
+        # handles primary methods and fields
+        elif (p := node.parent()) and (
+            isinstance(p, dyno.AggregateDecl) or isinstance(p, dyno.Interface)
+        ):
+            aggregate_name = get_single_name(p)
+            name = f"{aggregate_name}.{name}"
+
+        if (mod := get_module(node.parent())) and mod.kind() != "implicit":
```

Comment:
Unless the module is the direct parent of `node`, you don't want to include the name, right? Though in fact, you should be able to assert here that the parent is a module instead, since you require that in your pattern below.

Reply:
No, the way it prints is `Module.Aggregate.Symbol`, where `Module` and `Aggregate` are excluded if they don't exist. This was intentional, as when developing this I found it hard to at a glance identify what a symbol was